Support vars plugins in collections

s-hertel opened the pull request:
##### SUMMARY
Vars plugins were previously ignored in collections

This allows developers to require whitelisting for their vars plugin and users to control when vars plugins are executed.

The plugins themselves control whether or not whitelisting is required. Third-party plugins will continue to run by default unless they add whitelisting support. Since PluginLoader.all() does not search collections, vars plugins in collections will need to be in the list of enabled vars plugins. The `host_group_vars` plugin is now requires whitelisting and is whitelisted by default.

There are two new settings to control when vars plugins run and they default to the current behavior. They can be used in tandem or separately. RUN_VARS_PLUGINS is a global setting that determines whether to run vars plugins anytime variables are 'demanded' by tasks, or just at the 'start' after importing an inventory source. The other setting 'stage' is vars plugin specific and can be set in ansible.cfg or an env var.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
Vars plugins

bcoca committed:
draft

bcoca committed:
move var plugins hanlding to own file

	- add whitelisting
	- add 'stage' configuration ('all', 'task', 'inventory') for users to control when they execute.
	- TODO:
		- update ansible-inventory (when playbook dir) and InventoryManager to use them
		- update config, default to host_group_vars, update plugin to use 'stage' config

s-hertel committed:
Add whitelisting and enable host_group_vars by default

Update vars plugins to use config manager and add stage configuration

Update ansible-inventory to share code and use stage

Update host_group_vars to use stage

s-hertel committed:
Add tests

s-hertel committed:
fix bad copy pasta

ci_complete

s-hertel committed:
is_collection_ref doesn't exist anymore

ci_complete

bcoca commented on lib/ansible/plugins/vars/__init__.py:
```diff
@@ -18,22 +18,29 @@
+        super(BaseVarsPlugin, self).__init__()
         self._display = display
 
+    def has_option(self, option):
```
move this to AnsiblePlugin?

bcoca commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -28,6 +28,17 @@
         - Hidden (starting with '.') and backup (ending with '~') files and directories are ignored.
         - Only applies to inventory sources that are existing paths.
     options:
+      stage:
```
not a blocker, but is it worth making a doc fragment?

s-hertel committed:
Move has_option to AnsiblePlugin

Use a doc fragment for vars plugins

use vars_ sections

boilerplate

s-hertel committed:
document new req

s-hertel committed:
document that collections can use vars plugins

samccann commented on docs/docsite/rst/plugins/vars.rst:
```diff
@@ -19,16 +19,34 @@ The :ref:`host_group_vars <host_group_vars_vars>` plugin shipped with Ansible en
-You can activate a custom vars plugin by either dropping it into a ``vars_plugins`` directory adjacent to your play,  inside a role, or by putting it in one of the directory sources configured in :ref:`ansible.cfg <ansible_configuration_settings>`.
+You can activate a custom vars plugin by either dropping it into a ``vars_plugins`` directory adjacent to your play, inside a role, or by putting it in one of the directory sources configured in :ref:`ansible.cfg <ansible_configuration_settings>`.
 
+Starting in Ansible 2.10 vars plugins can require whitelisting rather than running by default. To enable a plugin that requires whitelisting set ``vars_plugins_enabled`` in the ``defaults`` section of :ref:`ansible.cfg <ansible_configuration_settings>` or set the ``ANSIBLE_VARS_ENABLED`` environment variable to the list of vars plugins you want to execute. By default the :ref:`host_group_vars <ho ...
```
```suggestion
Starting in Ansible 2.10, vars plugins can require whitelisting rather than running by default. To enable a plugin that requires whitelisting set ``vars_plugins_enabled`` in the ``defaults`` section of :ref:`ansible.cfg <ansible_configuration_settings>` or set the ``ANSIBLE_VARS_ENABLED`` environment variable to the list of vars plugins you want to execute. By default, the :ref:`host_group_vars <host_group_vars_vars>` plugin shipped with Ansible is whitelisted.
```

samccann commented on docs/docsite/rst/plugins/vars.rst:
```diff
@@ -19,16 +19,34 @@ The :ref:`host_group_vars <host_group_vars_vars>` plugin shipped with Ansible en
 
+Starting in Ansible 2.10 vars plugins can require whitelisting rather than running by default. To enable a plugin that requires whitelisting set ``vars_plugins_enabled`` in the ``defaults`` section of :ref:`ansible.cfg <ansible_configuration_settings>` or set the ``ANSIBLE_VARS_ENABLED`` environment variable to the list of vars plugins you want to execute. By default the :ref:`host_group_vars <ho ...
+
+Starting in Ansible 2.10 vars plugins in collections can be used. All vars plugins in collections require whitelisting and need to use the fully qualified name in the format ``namespace.collection_name.vars_plugin_name``.
```
```suggestion
Starting in Ansible 2.10, you can use vars plugins in collections. All vars plugins in collections require whitelisting and need to use the fully qualified collection name in the format ``namespace.collection_name.vars_plugin_name``.
```

samccann commented on lib/ansible/config/base.yml:
```diff
@@ -1745,6 +1757,15 @@ USE_PERSISTENT_CONNECTIONS:
+  ini:
+  - {key: vars_plugins_enabled, section: defaults}
+  type: list
+  version_added: "2.9"
```
was this really added in 2.9 or did you mean 2.10 here?

s-hertel replied:
It was 2.9 at the time of opening the PR. I need to update these.

samccann commented on lib/ansible/plugins/doc_fragments/vars_plugin_staging.py:
```diff
@@ -0,0 +1,20 @@
+    description: Control when this vars plugin may be executed.
+    choices: ['all', 'task', 'inventory']
+    default: 'all'
+    version_added: "2.9"
```
2.9 or 2.10??

samccann commented on lib/ansible/plugins/doc_fragments/vars_plugin_staging.py:
```diff
@@ -0,0 +1,20 @@
+options:
+  stage:
+    description: Control when this vars plugin may be executed.
+    choices: ['all', 'task', 'inventory']
```
can you describe what the different choices mean here?

s-hertel replied:
Added in c2bdd51

samccann commented on test/integration/targets/collections/collection_root_user/ansible_collections/testns/testcoll/plugins/vars/custom_vars.py:
```diff
@@ -0,0 +1,45 @@
+
+DOCUMENTATION = '''
+    vars: custom_vars
+    version_added: "2.9"
```
2.9 or 2.10

samccann commented on test/integration/targets/collections/collections/ansible_collections/testns/content_adj/plugins/vars/custom_adj_vars.py:
```diff
@@ -0,0 +1,45 @@
+
+DOCUMENTATION = '''
+    vars: custom_adj_vars
+    version_added: "2.9"
```
2.9 vs 2.10

samccann commented on test/integration/targets/collections/custom_vars_plugins/other_vars_plugin.py:
```diff
@@ -0,0 +1,45 @@
+
+DOCUMENTATION = '''
+    vars: other_vars_plugin
+    version_added: "2.9"
```
2.9 vs 2.10

samccann commented on test/integration/targets/collections/custom_vars_plugins/vars_req_whitelist.py:
```diff
@@ -0,0 +1,47 @@
+
+DOCUMENTATION = '''
+    vars: vars_req_whitelist
+    version_added: "2.9"
```
2.9 vs 2.10

samccann commented:
This needs a porting guide and changelog fragment as well.

bcoca commented:
@samccann  not really, it doesn't require changes to existing usages of vars plugins so I don't think a porting guide is required. As for the fragment, i agree, it should be a minor_change to be reflected in the changelog.

samccann commented:
Thanks @bcoca

s-hertel committed:
Apply suggestions from code review

Co-Authored-By: Sandra McCann <samccann@redhat.com>

s-hertel committed:
Move assessment of the global vars plugin toggle to vars/plugins.py

Change the per-vars-plugin stage default to None so the global setting can be used if per-plugin has not been specified

This retains backward compatibility when v1 vars plugins incorporate the stage setting

Add tests for that

s-hertel committed:
Update version_added

s-hertel committed:
Add better documentation for the vars plugin 'stage' option

s-hertel committed:
changelog

s-hertel committed:
Missed one

samccann commented on lib/ansible/config/base.yml:
```diff
@@ -1691,6 +1691,19 @@ RETRY_FILES_SAVE_PATH:
+  default: demand
+  description:
+    - This setting can be used to optimize vars_plugin usage depending on user's inventory size and play selection.
+    - Setting to 'demand' will run vars_plugins relative to inventory sources anytime vars are 'demanded' by tasks.
```
```suggestion
    - Setting to C(demand) will run vars_plugins relative to inventory sources anytime vars are 'demanded' by tasks.
```

s-hertel replied:
Going to let my ci_complete tests finish and then I'll update these. Thank you for re-reviewing!

samccann commented on lib/ansible/config/base.yml:
```diff
@@ -1691,6 +1691,19 @@ RETRY_FILES_SAVE_PATH:
+  description:
+    - This setting can be used to optimize vars_plugin usage depending on user's inventory size and play selection.
+    - Setting to 'demand' will run vars_plugins relative to inventory sources anytime vars are 'demanded' by tasks.
+    - Setting to 'start' will run vars_plugins relative to inventory sources after importing that inventory source.
```
```suggestion
    - Setting to C(start) will run vars_plugins relative to inventory sources after importing that inventory source.
```

s-hertel committed:
ci_complete

s-hertel committed:
Apply suggestions from code review

Co-Authored-By: Sandra McCann <samccann@redhat.com>
